signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,256 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implement the FileBufferedSyncedCollection class.
+
+The FileBufferedSyncedCollection is a concrete implementation of the buffer
+protocol established by the BufferedSyncedCollection class. It uses an
+in-memory cache to store data when in buffered mode. It is suitable for
+use with any file-based backend because it performs integrity checks based on
+whether or not the underlying file has been modified while buffering was
+activated.
+"""
+
+import errno
+import hashlib
+import json
+import os
+import sys
+
+from typing import Dict
+
+from .buffered_synced_collection import BufferedCollection
+from .caching import get_cache
+from .errors import MetadataError
+
+
+class FileBufferedCollection(BufferedCollection):
+    """Implement buffering for SyncedCollections with file-based backends.
+
+    All file-based backends can use the same set of integrity checks prior to a
+    buffer flush. This class standardizes that protocol.
+    """
+
+    _cache = get_cache()
+    _cached_collections: Dict[int, BufferedCollection] = {}
+    _BUFFER_CAPACITY = 32 * 2 ** 20  # 32 MB
+    _CURRENT_BUFFER_SIZE = 0
+
+    def __init__(self, filename, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._filename = filename
+
+    @classmethod
+    def get_buffer_capacity(cls):
+        """Get the current buffer capacity."""
+        return cls._BUFFER_CAPACITY
+
+    @classmethod
+    def set_buffer_capacity(cls, new_capacity):
+        """Update the buffer capacity."""
+        cls._BUFFER_CAPACITY = new_capacity
+
+    @classmethod
+    def get_current_buffer_size(cls):
+        """Get the total amount of data currently stored in the buffer."""
+        return FileBufferedCollection._CURRENT_BUFFER_SIZE
+
+    @staticmethod
+    def _hash(blob):
+        """Calculate and return the md5 hash value for the file data.
+
+        Parameters
+        ----------
+        blob : bytes
+            Byte literal to be hashed.
+
+        Returns
+        -------
+        str
+            The md5 hash of the input bytes.
+        """
+        if blob is not None:
+            m = hashlib.md5()
+            m.update(blob)
+            return m.hexdigest()
+
+    def _get_file_metadata(self):
+        """Return metadata of file.
+
+        Returns
+        -------
+        Tuple[int, float]
+            The size and last modification time of the associated file.
+        """
+        try:
+            metadata = os.stat(self._filename)
+            return metadata.st_size, metadata.st_mtime
+        except OSError as error:
+            if error.errno != errno.ENOENT:
```

Comment:
So this should implicitly return `None` if the error is `ENOENT`? Just want to be sure we're handling edge cases appropriately. That's not a documented return value and it could cause problems in the calling code?

Reply:
Good point, I'll document this return value. It does make sense to do this; imagine that an empty collection is initialized and then buffered mode is enabled. At that point, no file exists, so we need a way to identify a problem if the file is created by something else while the collection is in buffered mode.

Reply:
Sounds good. I hadn't thought of that case. This can be resolved if/when the changes have been made.